File: android/app/src/main/java/org/rhythmeta/maimaid/core/data/OtogameSheetMatcher.kt
package org.rhythmeta.maimaid.core.data

import java.text.Normalizer
import java.util.Locale
import org.rhythmeta.maimaid.core.database.SheetEntity
import org.rhythmeta.maimaid.core.database.SongEntity

internal class OtogameSheetMatcher(
    songs: List<SongEntity>,
    sheets: List<SheetEntity>,
) {
    private val songsById = songs.associateBy(SongEntity::songIdentifier)
    private val japaneseSheets = sheets.filter { it.regionJp && !it.isRemoved }

    fun match(playlog: OtogamePlaylog): SheetEntity? {
        val difficulty = OtogameImportPolicy.difficulty(OtogameImportPolicy.difficultyCode(playlog)) ?: return null
        val chartType = OtogameImportPolicy.chartType(playlog)
        val eligible = japaneseSheets.filter { sheet ->
            typeMatches(sheet, chartType) && difficultyMatches(sheet, difficulty, playlog.music.utageKanjiName)
        }

        val normalizedTitle = normalizeTitle(playlog.music.name)
        if (normalizedTitle.isEmpty()) return null
        return eligible.filter { sheet ->
            normalizeTitle(songsById[sheet.songIdentifier]?.title.orEmpty()) == normalizedTitle
        }.distinctBy(SheetEntity::sheetKey).singleOrNull()
    }

    private fun typeMatches(sheet: SheetEntity, chartType: String): Boolean =
        canonicalChartType(sheet.type) == canonicalChartType(chartType)

    private fun canonicalChartType(value: String): String = when (value.trim().lowercase(Locale.ROOT)) {
        "standard", "std", "sd" -> "std"
        "deluxe", "dx" -> "dx"
        "utage" -> "utage"
        else -> value.trim().lowercase(Locale.ROOT)
    }

    private fun difficultyMatches(sheet: SheetEntity, difficulty: String, utageKanji: String?): Boolean {
        if (difficulty != "utage") return sheet.difficulty.equals(difficulty, ignoreCase = true)
        val expectedKanji = normalizeUtageKanji(utageKanji.orEmpty())
        return expectedKanji.isEmpty() || normalizeUtageKanji(sheet.difficulty) == expectedKanji
    }

    private fun normalizeTitle(value: String): String = Normalizer.normalize(value, Normalizer.Form.NFKC)
        .replace(UtageTitlePrefix, "")
        .lowercase(Locale.ROOT)
        .filter(Char::isLetterOrDigit)

    private fun normalizeUtageKanji(value: String): String = Normalizer.normalize(value, Normalizer.Form.NFKC)
        .filter(Char::isLetterOrDigit)

    private companion object {
        val UtageTitlePrefix = Regex("^(?:\\[[^]]+]|【[^】]+】|［[^］]+］)\\s*")
    }
}

File: ios/maimaid/Services/OtogameSheetMatcher.swift
import Foundation

@MainActor
struct OtogameSheetMatcher {
    private let songsByIdentifier: [String: Song]
    private let japaneseSheets: [Sheet]

    init(songs: [Song], sheets: [Sheet]) {
        songsByIdentifier = Dictionary(songs.map { ($0.songIdentifier, $0) }, uniquingKeysWith: { first, _ in first })
        japaneseSheets = sheets.filter(\.regionJp)
    }

    func match(_ playlog: OtogamePlaylog) -> Sheet? {
        guard let difficulty = OtogameImportPolicy.difficulty(
            for: OtogameImportPolicy.difficultyCode(for: playlog)
        ) else {
            return nil
        }

        let expectedType = OtogameImportPolicy.chartType(for: playlog)
        let expectedTitle = normalizeTitle(playlog.music.name)
        guard !expectedTitle.isEmpty else {
            return nil
        }

        let matches = japaneseSheets.filter { sheet in
            guard chartType(sheet.type) == expectedType,
                  difficultyMatches(sheet, expected: difficulty, utageKanji: playlog.music.utageKanjiName) else {
                return false
            }
            let title = sheet.song?.title ?? songsByIdentifier[sheet.songIdentifier]?.title ?? ""
            return normalizeTitle(title) == expectedTitle
        }
        return matches.count == 1 ? matches[0] : nil
    }

    func match(_ entry: OtogameRatingEntry) -> Sheet? {
        match(
            music: entry.music,
            difficultyCode: entry.levelInfo.difficulty
        )
    }

    private func match(music: OtogameMusic, difficultyCode: Int) -> Sheet? {
        guard let difficulty = OtogameImportPolicy.difficulty(for: difficultyCode) else {
            return nil
        }
        let expectedType = music.isDeluxe ? "dx" : "standard"
        let expectedTitle = normalizeTitle(music.name)
        guard !expectedTitle.isEmpty else {
            return nil
        }
        let matches = japaneseSheets.filter { sheet in
            guard chartType(sheet.type) == expectedType,
                  difficultyMatches(sheet, expected: difficulty, utageKanji: music.utageKanjiName) else {
                return false
            }
            let title = sheet.song?.title ?? songsByIdentifier[sheet.songIdentifier]?.title ?? ""
            return normalizeTitle(title) == expectedTitle
        }
        return matches.count == 1 ? matches[0] : nil
    }

    private func chartType(_ value: String) -> String {
        switch value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased() {
        case "std", "standard": "standard"
        case "utage": "utage"
        default: value.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        }
    }

    private func difficultyMatches(_ sheet: Sheet, expected: String, utageKanji: String?) -> Bool {
        if expected != "utage" {
            return sheet.difficulty.localizedCaseInsensitiveCompare(expected) == .orderedSame
        }
        let expectedKanji = normalizeUtageKanji(utageKanji ?? "")
        return expectedKanji.isEmpty || normalizeUtageKanji(sheet.difficulty) == expectedKanji
    }

    private func normalizeTitle(_ value: String) -> String {
        let compatible = value.precomposedStringWithCompatibilityMapping
            .trimmingCharacters(in: .whitespacesAndNewlines)
        let withoutPrefix = removingUtagePrefix(from: compatible)
        return withoutPrefix.lowercased().unicodeScalars.reduce(into: "") { result, scalar in
            if scalar.properties.isAlphabetic || scalar.properties.numericType != nil {
                result.unicodeScalars.append(scalar)
            }
        }
    }

    private func normalizeUtageKanji(_ value: String) -> String {
        value.precomposedStringWithCompatibilityMapping.unicodeScalars.reduce(into: "") { result, scalar in
            if scalar.properties.isAlphabetic || scalar.properties.numericType != nil {
                result.unicodeScalars.append(scalar)
            }
        }
    }

    private func removingUtagePrefix(from value: String) -> String {
        guard let first = value.first else {
            return value
        }
        let closingCharacter: Character? = switch first {
        case "[": "]"
        case "【": "】"
        case "［": "］"
        default: nil
        }
        guard let closingCharacter,
              let closingIndex = value.firstIndex(of: closingCharacter) else {
            return value
        }
        return String(value[value.index(after: closingIndex)...])
            .trimmingCharacters(in: .whitespacesAndNewlines)
    }
}
